File: 11/gpt-5.1/python/solution.py
from __future__ import annotations

from collections import defaultdict
from typing import Dict, List


def parse_graph(text: str) -> Dict[str, List[str]]:
    graph: Dict[str, List[str]] = {}
    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line:
            continue

        parts = line.split(":", 1)
        if len(parts) != 2:
            continue

        src = parts[0].strip()
        targets_part = parts[1].strip()

        targets: List[str] = []
        if targets_part:
            for t in targets_part.split():
                tt = t.strip()
                if tt:
                    targets.append(tt)

        graph[src] = targets

    return graph


def count_paths(node: str, destination: str, graph: Dict[str, List[str]], memo: Dict[str, int]) -> int:
    if node == destination:
        return 1

    if node in memo:
        return memo[node]

    total = 0
    for child in graph.get(node, []):
        total += count_paths(child, destination, graph, memo)

    memo[node] = total
    return total


def count_paths_with_required_nodes(
    node: str,
    destination: str,
    graph: Dict[str, List[str]],
    memo: Dict[tuple[str, int], int],
    mask: int,
) -> int:
    updated_mask = mask
    if node == "fft":
        updated_mask |= 1
    elif node == "dac":
        updated_mask |= 2

    if node == destination:
        return 1 if (updated_mask & 0b11) == 0b11 else 0

    key = (node, updated_mask)
    if key in memo:
        return memo[key]

    total = 0
    for child in graph.get(node, []):
        total += count_paths_with_required_nodes(child, destination, graph, memo, updated_mask)

    memo[key] = total
    return total


def run_example_test() -> None:
    example = """
aaa: you hhh
you: bbb ccc
bbb: ddd eee
ccc: ddd eee fff
ddd: ggg
eee: out
fff: out
ggg: out
hhh: ccc fff iii
iii: out
""".strip()

    graph = parse_graph(example)
    memo: Dict[str, int] = {}
    paths = count_paths("you", "out", graph, memo)

    if paths != 5:
        print(f"Example test failed: expected 5, got {paths}", flush=True)


def run_part2_example_test() -> None:
    example = """
svr: aaa bbb
aaa: fft
fft: ccc
bbb: tty
tty: ccc
ccc: ddd eee
ddd: hub
hub: fff
eee: dac
dac: fff
fff: ggg hhh
ggg: out
hhh: out
""".strip()

    graph = parse_graph(example)
    memo: Dict[tuple[str, int], int] = {}
    paths = count_paths_with_required_nodes("svr", "out", graph, memo, 0)

    if paths != 2:
        print(f"Part 2 example test failed: expected 2, got {paths}", flush=True)


def main() -> None:
    run_example_test()
    run_part2_example_test()

    with open("input.txt", "r", encoding="utf-8") as f:
        input_data = f.read().strip()

    graph_real = parse_graph(input_data)
    memo_real: Dict[str, int] = {}
    part1 = count_paths("you", "out", graph_real, memo_real)
    print(f"Day 11 Part 1: {part1}")

    memo_part2: Dict[tuple[str, int], int] = {}
    part2 = count_paths_with_required_nodes("svr", "out", graph_real, memo_part2, 0)
    print(f"Day 11 Part 2: {part2}")


if __name__ == "__main__":
    main()

File: 11/gpt-5.1/typescript/solution.ts
import * as fs from 'fs';

function parseGraph(text: string): Map<string, string[]> {
	const graph = new Map<string, string[]>();

	for (const rawLine of text.split(/\r?\n/)) {
		const line = rawLine.trim();
		if (!line) continue;

		const parts = line.split(':', 2);
		if (parts.length !== 2) continue;

		const from = parts[0].trim();
		const targetsPart = parts[1].trim();

		const targets: string[] = [];
		if (targetsPart) {
			for (const t of targetsPart.split(/\s+/)) {
				const tt = t.trim();
				if (tt) targets.push(tt);
			}
		}

		graph.set(from, targets);
	}

	return graph;
}

function countPaths(
	node: string,
	destination: string,
	graph: Map<string, string[]>,
	memo: Map<string, number>
): number {
	if (node === destination) return 1;

	const cached = memo.get(node);
	if (cached !== undefined) return cached;

	let total = 0;
	const children = graph.get(node) ?? [];
	for (const child of children) {
		total += countPaths(child, destination, graph, memo);
	}

	memo.set(node, total);
	return total;
}

function countPathsWithRequiredNodes(
	node: string,
	destination: string,
	graph: Map<string, string[]>,
	memo: Map<string, number>,
	mask: number,
): number {
	let updatedMask = mask;
	if (node === 'fft') {
		updatedMask |= 1;
	} else if (node === 'dac') {
		updatedMask |= 2;
	}

	if (node === destination) {
		return (updatedMask & 0b11) === 0b11 ? 1 : 0;
	}

	const key = `${node}|${updatedMask}`;
	const cached = memo.get(key);
	if (cached !== undefined) {
		return cached;
	}

	let total = 0;
	const children = graph.get(node) ?? [];
	for (const child of children) {
		total += countPathsWithRequiredNodes(child, destination, graph, memo, updatedMask);
	}

	memo.set(key, total);
	return total;
}

function runExampleTest(): void {
	const example = `
aaa: you hhh
you: bbb ccc
bbb: ddd eee
ccc: ddd eee fff
ddd: ggg
eee: out
fff: out
ggg: out
hhh: ccc fff iii
iii: out
`.trim();

	const graph = parseGraph(example);
	const memo = new Map<string, number>();
	const paths = countPaths('you', 'out', graph, memo);

	if (paths !== 5) {
		console.error(`Example test failed: expected 5, got ${paths}`);
	}
}

function runPart2ExampleTest(): void {
	const example = `
svr: aaa bbb
aaa: fft
fft: ccc
bbb: tty
tty: ccc
ccc: ddd eee
ddd: hub
hub: fff
eee: dac
dac: fff
fff: ggg hhh
ggg: out
hhh: out
`.trim();

	const graph = parseGraph(example);
	const memo = new Map<string, number>();
	const paths = countPathsWithRequiredNodes('svr', 'out', graph, memo, 0);

	if (paths !== 2) {
		console.error(`Part 2 example test failed: expected 2, got ${paths}`);
	}
}

runExampleTest();
runPart2ExampleTest();

const input = fs.readFileSync('input.txt', 'utf-8').trim();
const graphReal = parseGraph(input);
const memoReal = new Map<string, number>();

const part1 = countPaths('you', 'out', graphReal, memoReal);
console.log(`Day 11 Part 1: ${part1}`);

const memoPart2 = new Map<string, number>();
const part2 = countPathsWithRequiredNodes('svr', 'out', graphReal, memoPart2, 0);
console.log(`Day 11 Part 2: ${part2}`);
